runtime/src/iree/hal/drivers/hip/hip_device.c

```diff
@@ -1174,23 +1029,400 @@ static iree_status_t iree_hal_hip_device_queue_write(
+                        device->topology.devices[idx].hip_dispatch_stream,
+                        iree_hal_hip_event_handle(event), 0)));
+    iree_hal_hip_event_release(event);
+    IREE_TRACE_ZONE_END(z1);
+  }
+  IREE_TRACE_ZONE_END(z2);
+  if (!iree_status_is_ok(status)) {
+    IREE_TRACE_ZONE_END(z0);
+    return status;
+  }
+
+  // We have satisfied all of the waits.
+
+  IREE_TRACE_ZONE_BEGIN_NAMED(z3, "Launch");
+  iree_hal_command_buffer_t* command_buffer = data->command_buffer;
+  if (iree_hal_hip_multi_queue_command_buffer_isa(command_buffer)) {
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, iree_hal_hip_multi_queue_command_buffer_get(
+                command_buffer, data->queue_affinity, &command_buffer));
+  }
+  iree_hal_buffer_binding_table_t binding_table = data->binding_table;
+  if (iree_hal_deferred_command_buffer_isa(command_buffer)) {
+    iree_hal_command_buffer_t* stream_command_buffer = NULL;
+    iree_hal_command_buffer_mode_t mode =
+        iree_hal_command_buffer_mode(command_buffer) |
+        IREE_HAL_COMMAND_BUFFER_MODE_ONE_SHOT |
+        // NOTE: we need to validate if a binding table is provided as the
+        // bindings were not known when it was originally recorded.
+        (iree_hal_buffer_binding_table_is_empty(binding_table)
+             ? IREE_HAL_COMMAND_BUFFER_MODE_UNVALIDATED
+             : 0);
+    iree_hal_hip_device_create_stream_command_buffer(
+        (iree_hal_device_t*)data->device, mode,
+        command_buffer->allowed_categories, data->queue_affinity, 0,
+        &stream_command_buffer);
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, iree_hal_resource_set_insert(data->resource_set, 1,
+                                         &stream_command_buffer));
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, iree_hal_deferred_command_buffer_apply(
+                command_buffer, stream_command_buffer, binding_table));
+    data->command_buffer = stream_command_buffer;
+  } else if (iree_hal_hip_stream_command_buffer_isa(command_buffer)) {
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0,
+        iree_hal_resource_set_insert(data->resource_set, 1, &command_buffer));
+  } else if (iree_hal_hip_graph_command_buffer_isa(command_buffer)) {
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0,
+        iree_hal_resource_set_insert(data->resource_set, 1, &command_buffer));
+    hipGraphExec_t exec =
+        iree_hal_hip_graph_command_buffer_handle(command_buffer);
+    IREE_TRACE_ZONE_BEGIN_NAMED(z3, "HipGraphLaunch");
+    status = IREE_HIP_RESULT_TO_STATUS(
+        data->device->hip_symbols,
+        hipGraphLaunch(exec,
+                       device->topology.devices[idx].hip_dispatch_stream));
+    IREE_TRACE_ZONE_END(z3);
+  } else {
+    status = iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                              "unsupported command buffer type");
+  }
+
+  IREE_TRACE_ZONE_END(z3);
+
+  IREE_TRACE_ZONE_BEGIN_NAMED(z4, "Record completion semaphores");
```

Comment:
these should really all be separate functions at this point - the number of zones and amount of control flow is far too high. You're also leaking statuses/failign to check them in many places that are too hard to track. I'm giving up for now.